File: python/algo/202502/algo0518.py
from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # 自行解答 完全背包 
        dp = [0] * (amount + 1)
        dp[0] = 1
        for c in coins:
            for k in range(c, amount+1):
                dp[k] += dp[k-c]
        return dp[-1]

File: cplus/leetcode/202502/algo0518.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 -  完全背包
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        // 自行解答 完全背包 
        vector<unsigned> dp(amount+1, 0); // unsigned 不会溢出
        dp[0] = 1;
        for (auto& x : coins) 
            for (int k=x; k<=amount; k++) 
                dp[k] += dp[k-x];  

        return dp[amount];       
    }
};